File: load-test.js
#!/usr/bin/env node
//
// Load test for the trading exchange.
// Simulates N traders connecting via WebSocket, placing orders, and cancelling.
//
// Usage:
//   npm install ws
//   node load-test.js                          # 60 traders against localhost
//   node load-test.js --url wss://exchange.columbia.trade --traders 60
//   node load-test.js --url ws://localhost:8080 --traders 30 --ramp-delay 50
//
// Flags:
//   --url <ws url>          WebSocket base URL (default: ws://localhost:8080)
//   --traders <n>           Number of traders to simulate (default: 60)
//   --ramp-delay <ms>       Delay between each trader connecting (default: 100)
//   --order-interval <ms>   Base interval between orders per trader (default: 1500)
//   --duration <s>          How long to run after all traders connect (default: 120)
//   --market-data           Also open a market_data/ws connection per trader
//   --no-trade              Connect only, don't send orders (test connection stability)

const WebSocket = require("ws");

// ---------------------------------------------------------------------------
// CLI args
// ---------------------------------------------------------------------------
const args = process.argv.slice(2);
function flag(name, fallback) {
  const i = args.indexOf(`--${name}`);
  if (i === -1) return fallback;
  return args[i + 1];
}
function hasFlag(name) {
  return args.includes(`--${name}`);
}

const BASE_URL = flag("url", "ws://localhost:8080");
const NUM_TRADERS = parseInt(flag("traders", "60"), 10);
const RAMP_DELAY = parseInt(flag("ramp-delay", "100"), 10);
const ORDER_INTERVAL = parseInt(flag("order-interval", "1500"), 10);
const DURATION = parseInt(flag("duration", "120"), 10);
const OPEN_MARKET_DATA = hasFlag("market-data");
const NO_TRADE = hasFlag("no-trade");

const ASSETS = ["AD", "TS", "TT"];

// ---------------------------------------------------------------------------
// Stats
// ---------------------------------------------------------------------------
const stats = {
  connected: 0,
  disconnected: 0,
  ordersSent: 0,
  ordersConfirmed: 0,
  ordersRejected: 0,
  cancelsSent: 0,
  tradesBroadcast: 0,
  errors: 0,
  marketDataConnected: 0,
  marketDataDisconnected: 0,
};

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------
function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

function passwordArray(pw) {
  return pw.split("");
}

function log(traderId, msg) {
  const ts = new Date().toISOString().slice(11, 23);
  console.log(`[${ts}] ${traderId}: ${msg}`);
}

// ---------------------------------------------------------------------------
// Trader simulation
// ---------------------------------------------------------------------------
async function simulateTrader(traderId, password) {
  const url = `${BASE_URL}/orders/ws`;
  const protocol = `${traderId}|${password}`;
  const pwArray = passwordArray(password);

  let ws;
  try {
    ws = new WebSocket(url, [protocol]);
  } catch (err) {
    log(traderId, `FAILED TO CREATE WS: ${err.message}`);
    stats.errors++;
    return;
  }

  const activeOrders = []; // track order IDs for cancellation
  let gameStarted = false;

  return new Promise((resolve) => {
    let orderLoop = null;

    ws.on("open", () => {
      stats.connected++;
      log(traderId, `connected (${stats.connected}/${NUM_TRADERS})`);
    });

    ws.on("message", (raw) => {
      let msg;
      try {
        msg = JSON.parse(raw);
      } catch {
        return;
      }

      if ("GameStartedMessage" in msg) {
        gameStarted = true;
        log(traderId, "game started");
      }

      if ("OrderConfirmMessage" in msg) {
        stats.ordersConfirmed++;
        const orderId = msg.OrderConfirmMessage.order_info.order_id;
        activeOrders.push(msg.OrderConfirmMessage.order_info);
      }

      if ("OrderPlaceErrorMessage" in msg) {
        stats.ordersRejected++;
      }

      if ("CancelConfirmMessage" in msg) {
        const cancelledId = msg.CancelConfirmMessage.order_info.order_id;
        const idx = activeOrders.findIndex((o) => o.order_id === cancelledId);
        if (idx !== -1) activeOrders.splice(idx, 1);
      }

      if ("TradeOccurredMessage" in msg) {
        stats.tradesBroadcast++;
      }

      if ("Error" in msg) {
        stats.errors++;
        log(traderId, `ERROR: ${msg.Error}`);
      }
    });

    ws.on("close", (code, reason) => {
      stats.disconnected++;
      log(
        traderId,
        `DISCONNECTED code=${code} reason=${reason || "none"} (${stats.disconnected} total disconnects)`
      );
      if (orderLoop) clearInterval(orderLoop);
      resolve();
    });

    ws.on("error", (err) => {
      stats.errors++;
      log(traderId, `WS ERROR: ${err.message}`);
    });

    // Start sending orders after connection
    if (!NO_TRADE) {
      ws.on("open", () => {
        // Randomize interval per trader so they don't all fire at once
        const jitter = Math.floor(Math.random() * ORDER_INTERVAL);
        orderLoop = setInterval(() => {
          if (ws.readyState !== WebSocket.OPEN) return;

          // 20% chance to cancel an existing order if we have any
          if (activeOrders.length > 0 && Math.random() < 0.2) {
            const order =
              activeOrders[Math.floor(Math.random() * activeOrders.length)];
            const cancelMsg = {
              MessageType: "CancelRequest",
              OrderId: order.order_id,
              TraderId: traderId,
              Price: order.price,
              Symbol: order.symbol,
              Side: order.order_type,
              Password: pwArray,
            };
            ws.send(JSON.stringify(cancelMsg));
            stats.cancelsSent++;
            return;
          }

          // Place a new order
          const asset = ASSETS[Math.floor(Math.random() * ASSETS.length)];
          const side = Math.random() > 0.5 ? "Buy" : "Sell";
          const price = Math.floor(30 + Math.random() * 40); // 30-70 cents
          const amount = Math.floor(1 + Math.random() * 10); // 1-10 shares

          const orderMsg = {
            MessageType: "OrderRequest",
            OrderType: side,
            Amount: amount,
            Price: price,
            Symbol: asset,
            TraderId: traderId,
            Password: pwArray,
          };
          ws.send(JSON.stringify(orderMsg));
          stats.ordersSent++;
        }, ORDER_INTERVAL + jitter);
      });
    }

    // Auto-close after duration
    setTimeout(() => {
      if (ws.readyState === WebSocket.OPEN) {
        ws.close();
      }
      resolve();
    }, (DURATION + 30) * 1000); // extra 30s buffer for ramp-up
  });
}

// ---------------------------------------------------------------------------
// Market data connection (optional, tests broadcast fan-out)
// ---------------------------------------------------------------------------
function openMarketData(label) {
  const url = `${BASE_URL}/market_data/ws`;
  const ws = new WebSocket(url);
  ws.on("open", () => {
    stats.marketDataConnected++;
  });
  ws.on("close", () => {
    stats.marketDataDisconnected++;
    log(label, "MARKET DATA DISCONNECTED");
  });
  ws.on("error", (err) => {
    stats.errors++;
  });
  return ws;
}

// ---------------------------------------------------------------------------
// Stats printer
// ---------------------------------------------------------------------------
function printStats() {
  console.log("\n--- Load Test Stats ---");
  console.log(`  Connected:          ${stats.connected}/${NUM_TRADERS}`);
  console.log(`  Disconnected:       ${stats.disconnected}`);
  console.log(`  Orders sent:        ${stats.ordersSent}`);
  console.log(`  Orders confirmed:   ${stats.ordersConfirmed}`);
  console.log(`  Orders rejected:    ${stats.ordersRejected}`);
  console.log(`  Cancels sent:       ${stats.cancelsSent}`);
  console.log(`  Trades broadcast:   ${stats.tradesBroadcast}`);
  console.log(`  Errors:             ${stats.errors}`);
  if (OPEN_MARKET_DATA) {
    console.log(`  MD connected:       ${stats.marketDataConnected}`);
    console.log(`  MD disconnected:    ${stats.marketDataDisconnected}`);
  }
  console.log("------------------------\n");
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------
async function main() {
  console.log(`\nLoad test: ${NUM_TRADERS} traders → ${BASE_URL}`);
  console.log(
    `  ramp-delay=${RAMP_DELAY}ms  order-interval=${ORDER_INTERVAL}ms  duration=${DURATION}s`
  );
  console.log(
    `  market-data=${OPEN_MARKET_DATA}  no-trade=${NO_TRADE}\n`
  );

  const traders = [];
  const mdConnections = [];

  // Print stats every 10 seconds
  const statsInterval = setInterval(printStats, 10000);

  for (let i = 1; i <= NUM_TRADERS; i++) {
    const traderId = `trader${i}`;
    const password = String(i).padStart(4, "0");

    traders.push(simulateTrader(traderId, password));

    if (OPEN_MARKET_DATA) {
      mdConnections.push(openMarketData(`md-${traderId}`));
    }

    // Stagger connections
    if (i < NUM_TRADERS) await sleep(RAMP_DELAY);
  }

  console.log(`\nAll ${NUM_TRADERS} traders launched. Running for ${DURATION}s...\n`);

  // Wait for duration then wind down
  await sleep(DURATION * 1000);

  clearInterval(statsInterval);
  printStats();

  // Close any remaining market data connections
  for (const ws of mdConnections) {
    if (ws.readyState === WebSocket.OPEN) ws.close();
  }

  console.log("Load test complete.");
  process.exit(0);
}

main().catch((err) => {
  console.error("Fatal:", err);
  process.exit(1);
});

File: load_test.py
#!/usr/bin/env python3
"""
Load test for the trading exchange.
Simulates N traders connecting via WebSocket and placing random orders.

Usage:
    pip install websockets
    python load_test.py                          # default: 60 traders, localhost
    python load_test.py --url ws://exchange.columbia.edu --traders 60
"""

import asyncio
import json
import random
import time
import argparse
from collections import defaultdict

try:
    import websockets
except ImportError:
    print("Install websockets: pip install websockets")
    exit(1)

SYMBOLS = ["AD", "TS", "TT"]
STATS = defaultdict(int)
ERRORS = []
LATENCIES = []
CONNECTED = 0


async def trader_bot(trader_id: str, password: str, base_url: str, duration: int):
    global CONNECTED
    orders_url = f"{base_url}/orders/ws"
    subprotocol = f"{trader_id}|{password}"

    try:
        async with websockets.connect(
            orders_url,
            subprotocols=[subprotocol],
            open_timeout=10,
        ) as ws:
            CONNECTED += 1
            STATS["connections"] += 1
            print(f"  [{trader_id}] connected ({CONNECTED} total)")

            end_time = time.time() + duration
            order_count = 0

            while time.time() < end_time:
                symbol = random.choice(SYMBOLS)
                side = random.choice(["Buy", "Sell"])
                price = random.randint(20, 80)
                amount = random.randint(1, 10)

                msg = json.dumps({
                    "MessageType": "OrderRequest",
                    "Price": price,
                    "TraderId": trader_id,
                    "OrderType": side,
                    "Amount": amount,
                    "Password": list(password),
                    "Symbol": symbol,
                })

                t0 = time.time()
                await ws.send(msg)

                try:
                    resp = await asyncio.wait_for(ws.recv(), timeout=5)
                    latency_ms = (time.time() - t0) * 1000
                    LATENCIES.append(latency_ms)

                    data = json.loads(resp)
                    if "OrderConfirmMessage" in str(data):
                        STATS["order_placed"] += 1
                    elif "OrderFillMessage" in str(data) or "TradeOccurredMessage" in str(data):
                        STATS["fills"] += 1
                    elif "Error" in str(data):
                        STATS["order_errors"] += 1
                    else:
                        STATS["other_messages"] += 1
                except asyncio.TimeoutError:
                    STATS["timeouts"] += 1

                order_count += 1

                # Drain any extra broadcast messages (fills, etc.)
                try:
                    while True:
                        extra = await asyncio.wait_for(ws.recv(), timeout=0.05)
                        STATS["broadcast_msgs"] += 1
                except asyncio.TimeoutError:
                    pass

                # Random delay: simulate human-ish trading (0.5-2s between orders)
                await asyncio.sleep(random.uniform(0.5, 2.0))

            CONNECTED -= 1

    except Exception as e:
        STATS["connection_errors"] += 1
        ERRORS.append(f"[{trader_id}] {type(e).__name__}: {e}")


async def market_data_listener(base_url: str, duration: int):
    """Simulates a few market data subscribers (like the frontend chart)."""
    url = f"{base_url}/market_data/ws"
    try:
        async with websockets.connect(url, open_timeout=10) as ws:
            STATS["md_connections"] += 1
            end_time = time.time() + duration
            while time.time() < end_time:
                try:
                    msg = await asyncio.wait_for(ws.recv(), timeout=5)
                    STATS["md_messages"] += 1
                except asyncio.TimeoutError:
                    pass
    except Exception as e:
        ERRORS.append(f"[market_data] {type(e).__name__}: {e}")


async def run_load_test(base_url: str, num_traders: int, duration: int):
    print(f"Load test: {num_traders} traders, {duration}s duration")
    print(f"Target: {base_url}\n")

    # Build trader list (trader1..traderN with 4-digit passwords)
    traders = []
    for i in range(1, num_traders + 1):
        trader_id = f"trader{i}"
        password = f"{i:04d}"
        traders.append((trader_id, password))

    # Stagger connections over 5 seconds to avoid a thundering herd
    tasks = []
    for i, (tid, pw) in enumerate(traders):
        delay = (i / num_traders) * 5  # spread over 5s
        tasks.append(delayed_start(delay, trader_bot(tid, pw, base_url, duration)))

    # Add a few market data listeners
    for _ in range(5):
        tasks.append(market_data_listener(base_url, duration))

    print("Ramping up connections...")
    t_start = time.time()
    await asyncio.gather(*tasks, return_exceptions=True)
    elapsed = time.time() - t_start

    # Report
    print("\n" + "=" * 50)
    print(f"LOAD TEST RESULTS ({elapsed:.1f}s elapsed)")
    print("=" * 50)
    for k, v in sorted(STATS.items()):
        print(f"  {k:25s}: {v}")

    if LATENCIES:
        LATENCIES.sort()
        print(f"\n  {'latency (min)':25s}: {LATENCIES[0]:.1f} ms")
        print(f"  {'latency (median)':25s}: {LATENCIES[len(LATENCIES)//2]:.1f} ms")
        print(f"  {'latency (p95)':25s}: {LATENCIES[int(len(LATENCIES)*0.95)]:.1f} ms")
        print(f"  {'latency (p99)':25s}: {LATENCIES[int(len(LATENCIES)*0.99)]:.1f} ms")
        print(f"  {'latency (max)':25s}: {LATENCIES[-1]:.1f} ms")
        print(f"  {'throughput':25s}: {len(LATENCIES)/elapsed:.1f} orders/sec")

    if ERRORS:
        print(f"\nErrors ({len(ERRORS)}):")
        for e in ERRORS[:20]:
            print(f"  {e}")


async def delayed_start(delay: float, coro):
    await asyncio.sleep(delay)
    return await coro


def main():
    parser = argparse.ArgumentParser(description="Exchange load tester")
    parser.add_argument("--url", default="ws://localhost:8080",
                        help="WebSocket base URL (default: ws://localhost:8080)")
    parser.add_argument("--traders", type=int, default=60,
                        help="Number of simulated traders (default: 60)")
    parser.add_argument("--duration", type=int, default=30,
                        help="Test duration in seconds (default: 30)")
    args = parser.parse_args()

    asyncio.run(run_load_test(args.url, args.traders, args.duration))


if __name__ == "__main__":
    main()
